pkg/cloudprovider/controller/nodecontroller.go

```diff
@@ -475,14 +487,69 @@ func (nc NodeController) tryUpdateNodeStatus(node *api.Node) (error, time.Durati
+	// - saved status have no Ready Condition, but current one does - NodeController was restarter with Node data already present in etcd
+	// - saved status have some Ready Condition, but current one does not - it's an error, but we fill it up because that's probably a good thing to do
+	// - both saved and current statuses have Ready Conditions and they have the same LastProbeTime - nothing happened on that Node, it may be
+	//   unresponsive, so we leave it as it is
+	// - both saved and current statuses have Ready Conditions, they have different LastProbeTimes, but the same Ready Condition State -
+	//   everything's in order, we update probeTimestamp
+	// - both saved and current statuses have Ready Conditions, different LastProbeTimes and different Ready Condition State -
+	//   Ready Condition changed it state since we last seen it, so we update both probeTimestamp and readyTransitionTimestamp
+	// TODO: things to consider:
+	//   - if 'LastProbeTime' have gone back in time its probably and error, currently we ignore it
+	//   - currently only correct Ready State transition outside of Node Controller is marking it ready by Kubelet, we don't check
+	//     if that's the case, but it does not seem necessary.
+	if !found || (nc.getCondition(&storedNodeStatus.status, api.NodeReady) == nil && nc.getCondition(&node.Status, api.NodeReady) != nil) {
+		glog.Warningf("Observed status for new Node %s. Assuming now as a timestamp.", node.Name)
+		storedNodeStatus = NodeStatusData{
+			status:                   node.Status,
+			probeTimestamp:           nc.now(),
+			readyTransitionTimestamp: nc.now(),
+		}
+		nc.nodeStatusMap[node.Name] = storedNodeStatus
+	} else if nc.getCondition(&storedNodeStatus.status, api.NodeReady) != nil && nc.getCondition(&node.Status, api.NodeReady) == nil {
+		glog.Errorf("ReadyCondition was removed from Status of Node %s", node.Name)
+		// TODO: figure out what to do in this case. For now we do the same thing as above.
+		storedNodeStatus = NodeStatusData{
+			status:                   node.Status,
+			probeTimestamp:           nc.now(),
+			readyTransitionTimestamp: nc.now(),
+		}
+		nc.nodeStatusMap[node.Name] = storedNodeStatus
+	} else if nc.getCondition(&storedNodeStatus.status, api.NodeReady) != nil && nc.getCondition(&node.Status, api.NodeReady) != nil &&
+		nc.getCondition(&storedNodeStatus.status, api.NodeReady).LastProbeTime != nc.getCondition(&node.Status, api.NodeReady).LastProbeTime {
+		var transitionTime util.Time
+		if nc.getCondition(&storedNodeStatus.status, api.NodeReady).Status != nc.getCondition(&node.Status, api.NodeReady).Status {
+			transitionTime = nc.now()
+		} else {
+			transitionTime = storedNodeStatus.readyTransitionTimestamp
+		}
+		// If Status changes since the last time we checked, we update the timestamp to "now".
+		glog.V(3).Infof("Status has changed. Updating timestamp: %+v\n vs %+v.", storedNodeStatus.status, node.Status)
```

Comment:
Status changes is not the correct wording.  (partially due to the fact we have diferent 'status', as you've pointed out before).   What happened here is 'ready ondition updated' I think.